File: chat/uikit/lib/src/audio_recoder/audio_recorder.dart
import 'dart:async';

import 'package:flutter/material.dart';

import 'audio_recorder_platform.dart';

enum AudioRecordResultCode {
  successExceedMaxDuration(1),
  success(0),
  errorCancel(-1),
  errorRecording(-2),
  errorStorageUnavailable(-3),
  errorLessThanMinDuration(-4),
  errorRecordInnerFail(-5),
  errorRecordPermissionDenied(-6);

  final int code;
  const AudioRecordResultCode(this.code);

  static AudioRecordResultCode fromCode(int code) {
    return AudioRecordResultCode.values.firstWhere(
      (e) => e.code == code,
      orElse: () => AudioRecordResultCode.errorRecordInnerFail,
    );
  }
}

class RecordInfo {
  AudioRecordResultCode errorCode = AudioRecordResultCode.success;
  String path;
  final int duration;

  RecordInfo({
    required this.duration,
    required this.path,
  });

  @override
  String toString() {
    return 'RecordInfo(errorCode: ${errorCode.name}, path: $path, duration: $duration)';
  }
}

typedef RecordingProgressCallback = void Function(
    int duration, double progress);

typedef RecordingStateCallback = void Function(bool isRecording);

typedef RecordingPowerLevelCallback = void Function(int powerLevel);

typedef AudioRecordCompleteCallback = void Function(RecordInfo? recordInfo);

class AudioRecorder {
  bool _isRecording = false;
  final int maxDuration = 60000; // 1 minute
  final int minDuration = 1000; // 1 second

  RecordingProgressCallback? onProgressUpdate;
  RecordingStateCallback? onStateChanged;
  RecordingPowerLevelCallback? onPowerLevel;

  bool get isRecording => _isRecording;

  int _recordingDuration = 0;
  int get recordingDuration => _recordingDuration;

  double get recordingProgress => _recordingDuration / maxDuration;

  void initialize({
    RecordingProgressCallback? onProgressUpdate,
    RecordingStateCallback? onStateChanged,
    RecordingPowerLevelCallback? onPowerLevel,
  }) {
    this.onProgressUpdate = onProgressUpdate;
    this.onStateChanged = onStateChanged;
    this.onPowerLevel = onPowerLevel;
  }

  Future<bool> startRecord({
    required String filePath,
    required AudioRecordCompleteCallback onComplete,
    int? maxDurationMs,
  }) async {
    if (_isRecording) {
      debugPrint('Already recording');
      return false;
    }

    try {
      _recordingDuration = 0;
      _isRecording = true;

      onStateChanged?.call(_isRecording);

      // Set up callbacks for native platform
      AudioRecorderPlatform.setOnRecordTime((timeMs) {
        _recordingDuration = timeMs;
        onProgressUpdate?.call(_recordingDuration, recordingProgress);
      });

      AudioRecorderPlatform.setOnPowerLevel((powerLevel) {
        onPowerLevel?.call(powerLevel);
      });

      // Start native recording asynchronously
      _startNativeRecording(
        filePath: filePath,
        onComplete: onComplete,
        maxDurationMs: maxDurationMs ?? maxDuration,
      );

      return true;
    } catch (e) {
      debugPrint('Start record failed: $e');
      _cleanup();
      return false;
    }
  }

  Future<void> _startNativeRecording({
    required String filePath,
    required AudioRecordCompleteCallback onComplete,
    int? maxDurationMs,
  }) async {
    try {
      final result = await AudioRecorderPlatform.startRecordNative(
        config: AudioRecorderConfig(
          filepath: filePath,
          enableAIDeNoise: false,
          minDurationMs: minDuration,
          maxDurationMs: maxDurationMs ?? maxDuration,
        ),
      );

      RecordInfo? recordInfo;

      if (result.isSuccess && result.filePath != null) {
        recordInfo = RecordInfo(
          duration: (result.durationMs / 1000).floor(),
          path: result.filePath!,
        )..errorCode = result.resultCode;
      } else if (result.resultCode ==
          AudioRecordResultCode.errorLessThanMinDuration) {
        recordInfo = RecordInfo(
          duration: result.durationMs,
          path: '',
        )..errorCode = result.resultCode;
      } else {
        recordInfo = RecordInfo(
          duration: result.durationMs,
          path: result.filePath ?? '',
        )..errorCode = result.resultCode;
      }

      _cleanup();
      onComplete(recordInfo);
    } catch (e) {
      debugPrint('Native recording error: $e');
      _cleanup();
      onComplete(null);
    }
  }

  void stopRecord() {
    if (!_isRecording) {
      return;
    }

    try {
      AudioRecorderPlatform.stopRecordNative();
    } catch (e) {
      debugPrint('Stop record failed: $e');
    }
  }

  Future<void> cancelRecord() async {
    if (!_isRecording) {
      return;
    }

    try {
      await AudioRecorderPlatform.cancelRecordNative();
      _cleanup();
    } catch (e) {
      debugPrint('Cancel record failed: $e');
      _cleanup();
    }
  }

  void _cleanup() {
    final wasRecording = _isRecording;
    _isRecording = false;
    _recordingDuration = 0;

    if (wasRecording) {
      onStateChanged?.call(_isRecording);
      onProgressUpdate?.call(_recordingDuration, recordingProgress);
    }
  }

  void dispose() {
    _cleanup();
    AudioRecorderPlatform.dispose();
  }
}

File: chat/uikit/android/src/main/kotlin/com/tencent/cloud/tuikit/flutter/tuichatkit/audiorecorder/AudioRecorder.kt
package com.tencent.cloud.tuikit.flutter.tuichatkit.audiorecorder
import com.tencent.cloud.tuikit.flutter.tuichatkit.audiorecorder.audiorecorderimpl.AudioRecorderImpl

enum class ResultCode(val code: Int) {
    SUCCESS_EXCEED_MAX_DURATION(1),
    SUCCESS(0),
    ERROR_CANCEL(-1),
    ERROR_RECORDING(-2),
    ERROR_STORAGE_UNAVAILABLE(-3),
    ERROR_LESS_THAN_MIN_DURATION(-4),
    ERROR_RECORD_INNER_FAIL(-5),
    ERROR_RECORD_PERMISSION_DENIED(-6),
}

interface AudioRecorderListener {
    fun onCompleted(resultCode: ResultCode, path: String?, durationMs: Int)
}

object AudioRecorder {
    private val instance = AudioRecorderImpl()

    var onRecordTime: ((timeMs: Int) -> Unit)?
        get() = instance.onRecordTime
        set(value) { instance.onRecordTime = value }

    var onPowerLevel: ((powerLevel: Int) -> Unit)?
        get() = instance.onPowerLevel
        set(value) { instance.onPowerLevel = value }

    // To use AI noise reduction, the app must depend on LiteAVSDK_Professional v12.7+ and have the feature enabled.
    // Dependency: Add to app module's build.gradle dependencies: implementation("com.tencent.liteav:LiteAVSDK_Professional:12.7.0.xxxxx")
    // For enabling permissions, see documentation: https://cloud.tencent.com/document/product/269/113290
    //
    // 要使用 AI 降噪，应用必须依赖 LiteAVSDK_Professional v12.7+ 并启用该功能。依赖：在 app 模块的 build.gradle dependencies
    // 中添加：implementation("com.tencent.liteav:LiteAVSDK_Professional:12.7.0.xxxxx")
    fun startRecord(
        filepath: String? = null,
        enableAIDeNoise: Boolean = false,
        minDurationMs: Int = 1000,
        maxDurationMs: Int = 60000,
        listener: AudioRecorderListener
    ) {
        instance.startRecord(filepath, enableAIDeNoise, minDurationMs, maxDurationMs, listener)
    }

    fun stopRecord() {
        instance.stopRecord()
    }

    fun cancelRecord() {
        instance.cancelRecord()
    }
}
